### ISSUE TITLE
Compile iso2mesh for ARM Mac

### ISSUE THREAD

**Issue by Edouard2laire** [fangq/iso2mesh#74]
Iso2mesh needs to be recompiled for arm Mac.  I tried to recompile according to the instructions but I have an error when compiling JMeshLib.

See log file: 
https://justpaste.it/5o8b3

**Comment by Edouard2laire** [comment]
i made a bit of progress and was able to fix the error 

```/Users/edelaire1/Documents/software/iso2mesh/tools/meshfix/contrib/JMeshLib/include/clusterGraph.h:57:50: error: cast from pointer to smaller type 'j_voidint' (aka 'int') loses information
 clusterEdge *getFirst() {return (numels)?(edges[(j_voidint)getHead()]):(NULL);}```

by replaceing every cast to [(j_voidint) by (j_voidint)(size_t) as recomended here: https://stackoverflow.com/questions/22419063/error-cast-from-pointer-to-smaller-type-int-loses-information-in-eaglview-mm/22474752#22474752


I am now having an issue compiling OpenNL3.2.1


```/Users/edelaire1/Documents/software/iso2mesh/tools/meshfix/contrib/OpenNL3.2.1/src/nl_single_file.c:2505:10: error: call to undeclared function 'lsame_'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    if (!NL_FORTRAN_WRAP(lsame)(uplo, "U") && 
         ^
/Users/edelaire1/Documents/software/iso2mesh/tools/meshfix/contrib/OpenNL3.2.1/src/nl_single_file.c:420:28: note: expanded from macro 'NL_FORTRAN_WRAP'
#define NL_FORTRAN_WRAP(x) x##_
                           ^
<scratch space>:122:1: note: expanded from here
lsame_
^
/Users/edelaire1/Documents/software/iso2mesh/tools/meshfix/contrib/OpenNL3.2.1/src/nl_single_file.c:2526:2: error: call to undeclared function 'xerbla_'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
        NL_FORTRAN_WRAP(xerbla)("DTPSV ", &info);
        ^
/Users/edelaire1/Documents/software/iso2mesh/tools/meshfix/contrib/OpenNL3.2.1/src/nl_single_file.c:420:28: note: expanded from macro 'NL_FORTRAN_WRAP'
#define NL_FORTRAN_WRAP(x) x##_
                           ^
<scratch space>:129:1: note: expanded from here
xerbla_
^
/Users/edelaire1/Documents/software/iso2mesh/tools/meshfix/contrib/OpenNL3.2.1/src/nl_single_file.c:2367:22: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
/* Subroutine */ int NL_FORTRAN_WRAP(dtpsv)(uplo, trans, diag, n, ap, x, incx, uplo_len, 
                     ^
/Users/edelaire1/Documents/software/iso2mesh/tools/meshfix/contrib/OpenNL3.2.1/src/nl_single_file.c:420:28: note: expanded from macro 'NL_FORTRAN_WRAP'
#define NL_FORTRAN_WRAP(x) x##_
                           ^
<scratch space>:121:1: note: expanded from here
dtpsv_
^
/Users/edelaire1/Documents/software/iso2mesh/tools/meshfix/contrib/OpenNL3.2.1/src/nl_single_file.c:2779:82: warning: too few arguments in call to 'dtpsv_'
    NL_FORTRAN_WRAP(dtpsv)(UL[(int)uplo],T[(int)trans],D[(int)diag],&n,AP,x,&incx); 
    ~~~~~~~~~~~~~~~~~~~~~~                                                       ^
/Users/edelaire1/Documents/software/iso2mesh/tools/meshfix/contrib/OpenNL3.2.1/src/nl_single_file.c:3439:10: fatal error: 'slu_cdefs.h' file not found
#include <slu_cdefs.h>
         ^~~~~~~~~~~~~
2 warnings and 3 errors generated.
make[2]: *** [CMakeFiles/meshfix.dir/contrib/OpenNL3.2.1/src/nl_single_file.c.o] Error 1```

**Comment by Edouard2laire** [comment]
(cross-posting from https://github.com/fangq/meshfix/pull/2 - sorry, I am a bit confused with all the different repositories so I am not sure it's the right place)

I have now this error now when compiling cgalmesh: https://justpaste.it/3qrcl

**Comment by Edouard2laire** [comment]
Hi @fangq; 

Would you have any idea on how to solve that issue ? I am not very familiar with cmake and how linking works in general so I am not sure about what to search for to solve that error in the linking of the tbb library to compile cgalmesh.

**Comment by fangq** [comment]
@Edouard2laire, sorry that I haven't been paying much attention this thread - I saw there are multiple tickets, some have been resolved, some are not.

can you summarize what does this issue currently stand? is it just [cgalmesh compilation error](https://justpaste.it/3qrcl) or more related to components of meshfix?

for cgal, the upstream authors have been very responsive in their mailing list in the past, adding @lrineau to this thread.

**Comment by fangq** [comment]
@Edouard2laire, one thing I think you were facing was because the meshfix I linked as a submodule was an old version.

the newer version can be found at

https://github.com/MarcoAttene/MeshFix-V2.1

are you able to compile this one on ARM mac?

**Comment by Edouard2laire** [comment]
> @Edouard2laire, sorry that I haven't been paying much attention this thread - I saw there are multiple tickets, some have been resolved, some are not.

Yes, sorry. I got confused because of the different repo; and was not sure about where to post. 


> can you summarize what does this issue currently stand? is it just [cgalmesh compilation error](https://justpaste.it/3qrcl) or more related to components of meshfix?

Yes; now, the only issue is with cgalmesh. (https://justpaste.it/3qrcl). 


> for cgal, the upstream authors have been very responsive in their mailing list in the past, adding @lrineau to this thread.

Thank you. 


> @Edouard2laire, one thing I think you were facing was because the meshfix I linked as a submodule was an old version.

I will give a try and let you know. I was also able to compile using https://github.com/fangq/meshfix/pull/2

**Comment by Edouard2laire** [comment]
I might have found the solution here: https://oneapi-src.github.io/oneTBB/GSG/integrate.html#cmake with the following fix: 

``` 
create_single_source_cgal_program( "mesh_polyhedral_domain.cpp")
target_link_libraries(mesh_polyhedral_domain TBB::tbb) % NEW

 create_single_source_cgal_program( "mesh_3D_image.cpp")
 target_link_libraries(mesh_3D_image TBB::tbb)  % NEW

```

The program is now compiling

**Comment by Edouard2laire** [comment]
New question: The compilation is working well to generate the executables but it doesn't generate Mex files. Do you have any indication on how to generate those? 


Thanks a lot, 
Edouard

**Comment by Edouard2laire** [comment]
hello @fangq, 

Do you have any indication for the generation of the Mex files or should we ship only binaries? 

Thank you, 
Edouard

**Comment by fangq** [comment]
@Edouard2laire, all binaries under iso2mesh/bin folders are executables, they are not mex files, except that I named them with the mexext as suffix just to be able to store them under the same directory.

when I build the bundled executables, I usually use static linking in the link step so that the generated binary can be run on a wide range of machines - I also try to find a machine with older glibc version, say Ubuntu 16.04, to build the executable so that the binary can be launched in most newer machines.

**Comment by fangq** [comment]
closed with https://github.com/fangq/iso2mesh/commit/aa78f3a1c9ffdc9dfd41b1ae290161d3233110a5

**Comment by Edouard2laire** [comment]
thank you :)
